### ISSUE TITLE
Optimise Cluster VPN

### ISSUE THREAD

**Issue by vlerenc** [gardener/vpn#17]
# Story
* As provider I want to have a fast, reliable and safe cluster VPN, so that I get all these benefits for this crucial component in the seed & shoot approach.

# Motivation
Performance, simplicity, and security.

# Acceptance Criteria
- [ ] Investigate if there is a more elegant approach to SSH for the cluster VPN between the API server in the seed cluster and the workers in the shoot cluster
- [ ] Secure access with security groups

# Related Resources
* See https://groups.google.com/forum/#!topic/kubernetes-dev/rO-cUwJMixQ
* See https://groups.google.com/forum/#!msg/kubernetes-sig-api-machinery/bwJJ93qA99o/yejsv2JPBAAJ
* See [k8sniff](https://github.com/kubermatic/k8sniff)
* See [kubernikus wormhole](https://github.com/sapcc/kubernikus/tree/master/pkg/wormhole)

# Definition of Done
- [ ] Knowledge is distributed: Have you spread your knowledge in pair programming/code review?
- [ ] Unit Tests are provided: Have you written automated unit tests or added manual NGPTT tickets?
- [ ] Integration Tests are provided: Have you written automated integration tests?
- [ ] Minimum API exposure: If you have added public API, was it really necessary/is it minimal?
- [ ] Operations guide: Have you updated the [operations guide](https://github.wdf.sap.corp/kubernetes/kube-docs/wiki/Operations%20Guide)?

**Comment by marwinski** [comment]
OpenVPN could be used to replace the ssh solution, however only in TCP mode as AWS  (and probably other) load balancers only support TCP. With TCP mode however one of the key benefits compared to ssh is gone. 

The decision is to stay on ssh but attempt to fix the current problems:

1. server currently does not restrict the ciphers which can be used by the client -> configuration change
2. VPN from api server breaks down every 30 t o45 minutes and is restarted. This should not be the case according to AWS LB documentation (connection should remain open indefinitely)
3. VPN from prometheus breaks down every 3 minutes

(2) and (3) need debugging, (1) requires a configuration change and testing.

**Comment by marwinski** [comment]
Some further thoughts and results from today's investigation:

* problem is that ssh server actively closes the connection, error message in the debug2 log is always something like ```channel 0: adjust 98626 overflows remote window 4294869733```
* There is very little information on this on the web, mostly old stuff on this kind of error
* on the client side there are lots of messages like ``` channel 0: window 1989912 sent adjust 99700```. This might however be unrelated. 

This is not trivial to solve. Some suspicions before digging in to ssh:
1. We use the openssh that is part of the alpine packages. This does contain hpn ssh: https://www.psc.edu/hpn-ssh which does not appear to be widely used. There is also an interesting message at https://github.com/set-element/openssh-hpn-isshd stating that there is not enough time for the author to work on it. 
2. openssh used musl-libc on alpine. This might lead to some funny problems
3. the AWS ELB might interfere in some way however I have not found any hints that this might be the case

To validate certain assumptions I have built vpn containers based on debian  stable and the openssh implementation that it brings which is slightly older (7.4 vs 7.5). At first sight this does appear to fix the connection problems described above in (2) and (3). Results will be in tomorrow. 

While doing some research I found this very interesting site: https://turecki.net/content/getting-most-out-ssh-hardware-acceleration-tuning-aes-ni. This would be solve (1) above by pre-selecting a cipher that could be hardware accelerated.

Note: Problem not solved. This might point to suspicion 3 above (AWS ELB).

Next Options: 
- use OpenVPN over TCP
- deploy a dedicated VM to act as the ssh endpoint
